File: frontend/src/services/performance.ts
import request from './request';

export interface PerformanceTask {
  id: number;
  name: string;
  test_type: string;
  target_url: string;
  method: string;
  headers: Record<string, any>;
  body: Record<string, any>;
  business_volume: number;
  concurrency: number;
  duration_seconds: number;
  tps_target: number | null;
  ramp_up: number;
  script_type: string;
  script_content: string | null;
  jmeter_config: string | null;
  plan: any;
  status: string;
  error_message: string | null;
  created_at: string;
  updated_at: string;
}

export interface PerformanceResult {
  id: number;
  task_id: number;
  total_requests: number;
  total_errors: number;
  error_rate: number;
  avg_tps: number;
  peak_tps: number;
  avg_rt: number;
  p50_rt: number | null;
  p90_rt: number | null;
  p95_rt: number | null;
  p99_rt: number | null;
  concurrency: number;
  duration_seconds: number;
  analysis: any;
  status: string;
  created_at: string;
}

export interface PerformanceMetric {
  id: number;
  result_id: number;
  timestamp: number;
  elapsed: number;
  tps: number;
  avg_rt: number;
  concurrent_users: number;
  error_count: number;
  cpu_percent: number | null;
  memory_mb: number | null;
}

export function createTask(data: { name: string; target_url: string; method: string; test_type: string; headers?: any; body?: any; business_volume: number }): Promise<PerformanceTask> {
  return request.post('/performance', data);
}
export function listTasks(params: { page: number; page_size: number; test_type?: string; status?: string }): Promise<{ items: PerformanceTask[]; total: number; page: number; page_size: number } | PerformanceTask[]> {
  return request.get('/performance/list', { params });
}
export function getTask(taskId: number): Promise<PerformanceTask> {
  return request.get(`/performance/${taskId}`);
}
export function deleteTask(taskId: number): Promise<any> {
  return request.delete(`/performance/${taskId}`);
}
export function runPlan(taskId: number): Promise<any> {
  return request.post(`/performance/${taskId}/plan`);
}
export function runScript(taskId: number, scriptType: string): Promise<any> {
  return request.post(`/performance/${taskId}/script`, { script_type: scriptType });
}
export function runAnalysis(taskId: number, logs?: any[]): Promise<any> {
  return request.post(`/performance/${taskId}/analyze`, { logs: logs || null });
}
export function getResults(taskId: number): Promise<{ items: PerformanceResult[]; total: number } | PerformanceResult[]> {
  return request.get(`/performance/${taskId}/results`);
}
export function getMetrics(resultId: number): Promise<{ items: PerformanceMetric[]; total: number } | PerformanceMetric[]> {
  return request.get(`/performance/${resultId}/metrics`);
}
export function saveResult(taskId: number, data: any): Promise<any> {
  return request.post(`/performance/${taskId}/results`, data);
}
export function saveMetric(resultId: number, data: any): Promise<any> {
  return request.post(`/performance/${resultId}/metrics`, data);
}

// ================================================================
// 测试执行
// ================================================================

export interface ExecuteStatus {
  task_id: number;
  result_id: number;
  status: 'running' | 'completed' | 'failed' | 'stopped' | 'not_found';
  elapsed: number;
  error: string;
  metrics_count: number;
  final_stats: any;
}

export interface MetricEvent {
  type: 'metric' | 'final' | 'end' | 'stopped' | 'error';
  task_id?: number;
  result_id?: number;
  timestamp?: number;
  elapsed?: number;
  tps?: number;
  avg_rt?: number;
  concurrent_users?: number;
  error_count?: number;
  total_requests?: number;
  cpu_percent?: number | null;
  memory_mb?: number | null;
  p50_rt?: number;
  p90_rt?: number;
  p95_rt?: number;
  p99_rt?: number;
  min_rt?: number;
  max_rt?: number;
  status?: string;
  error?: string;
  data?: any;
  final_stats?: any;
  message?: string;
}

export function executeTask(taskId: number): Promise<{ status: string; task_id: number; result_id: number }> {
  return request.post(`/performance/${taskId}/execute`);
}

export function stopExecute(taskId: number): Promise<any> {
  return request.post(`/performance/${taskId}/execute/stop`);
}

export function getExecuteStatus(taskId: number): Promise<ExecuteStatus> {
  return request.get(`/performance/${taskId}/execute/status`);
}

/**
 * SSE 流式监听实时性能指标
 * 使用 fetch + ReadableStream 解析 text/event-stream
 *
 * @param taskId 任务 ID
 * @param onMetric 指标回调
 * @param onEnd 结束回调
 * @param onError 错误回调
 * @returns AbortController (调用 .abort() 停止监听)
 */
export function streamExecuteMetrics(
  taskId: number,
  onMetric: (event: MetricEvent) => void,
  onEnd?: (event: MetricEvent) => void,
  onError?: (error: Error) => void,
): AbortController {
  const controller = new AbortController();

  (async () => {
    try {
      const resp = await fetch(`/api/performance/${taskId}/execute/stream`, {
        method: 'GET',
        credentials: 'include',
        signal: controller.signal,
        headers: { 'Accept': 'text/event-stream' },
      });

      if (!resp.ok) {
        onError?.(new Error(`SSE 连接失败: ${resp.status}`));
        return;
      }

      const reader = resp.body?.getReader();
      if (!reader) {
        onError?.(new Error('无法读取响应流'));
        return;
      }

      const decoder = new TextDecoder();
      let buffer = '';

      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        buffer += decoder.decode(value, { stream: true });
        const lines = buffer.split('\n');
        buffer = lines.pop() || '';

        for (const line of lines) {
          if (line.startsWith('data: ')) {
            const jsonStr = line.slice(6).trim();
            if (!jsonStr) continue;
            try {
              const event: MetricEvent = JSON.parse(jsonStr);
              if (event.type === 'metric') {
                onMetric(event);
              } else if (event.type === 'final') {
                onMetric(event);
              } else if (event.type === 'end' || event.type === 'stopped') {
                onEnd?.(event);
                return;
              } else if (event.type === 'error') {
                onError?.(new Error(event.message || '执行错误'));
                return;
              }
            } catch {
              // 忽略解析错误
            }
          }
        }
      }
    } catch (err: any) {
      if (err.name !== 'AbortError') {
        onError?.(err);
      }
    }
  })();

  return controller;
}

File: backend/app/api/performance.py
"""
性能测试 API 路由

端点:
  任务管理:
    POST   /                          创建任务
    GET    /list                       任务列表
    GET    /{task_id}                  任务详情
    DELETE /{task_id}                  删除任务

  Agent 编排:
    POST   /{task_id}/plan             生成测试方案 (PerformancePlanAgent)
    POST   /{task_id}/script           生成脚本 (PerformanceScriptAgent)
    POST   /{task_id}/analyze          分析结果 (PerformanceAnalysisAgent)

  测试执行:
    POST   /{task_id}/execute          启动执行 (Locust 引擎)
    POST   /{task_id}/execute/stop     停止执行
    GET    /{task_id}/execute/status   执行状态
    GET    /{task_id}/execute/stream   实时指标 SSE 流

  结果和指标:
    GET    /{task_id}/results          执行结果列表
    GET    /{result_id}/metrics         实时指标
    POST   /{task_id}/results           手动保存结果
    POST   /{result_id}/metrics         手动保存指标
"""
import asyncio
import json
from typing import Any, Dict, List, Optional

from fastapi import APIRouter, Depends, Query
from fastapi.responses import StreamingResponse
from pydantic import BaseModel, Field

from app.core.auth import require_auth
from app.models.user import User
from app.schemas.response import Response
from app.services.performance_service import PerformanceService

router = APIRouter()
_service = PerformanceService()


# ================================================================
# Pydantic Schemas
# ================================================================

class TaskCreate(BaseModel):
    name: str = Field(..., description="任务名称")
    target_url: str = Field(..., description="目标接口 URL")
    method: str = Field(default="GET", description="HTTP 方法")
    test_type: str = Field(default="api", description="测试类型: api/web")
    headers: Optional[Dict[str, Any]] = Field(default=None, description="请求头")
    body: Optional[Dict[str, Any]] = Field(default=None, description="请求体")
    business_volume: int = Field(default=10000, description="预期日业务量")


class ScriptGenerate(BaseModel):
    script_type: str = Field(default="locust", description="脚本类型: locust/jmeter")


class ResultSave(BaseModel):
    total_requests: int = 0
    total_errors: int = 0
    error_rate: float = 0.0
    avg_tps: float = 0.0
    peak_tps: float = 0.0
    avg_rt: float = 0.0
    p50_rt: Optional[float] = None
    p90_rt: Optional[float] = None
    p95_rt: Optional[float] = None
    p99_rt: Optional[float] = None
    concurrency: int = 0
    duration_seconds: int = 0
    status: str = "completed"


class MetricSave(BaseModel):
    timestamp: float = 0.0
    elapsed: float = 0.0
    tps: float = 0.0
    avg_rt: float = 0.0
    concurrent_users: int = 0
    error_count: int = 0
    cpu_percent: Optional[float] = None
    memory_mb: Optional[float] = None


class AnalyzeRequest(BaseModel):
    logs: Optional[List[Any]] = Field(default=None, description="应用日志列表, 每条为 dict 或 str")


# ================================================================
# 任务管理
# ================================================================

@router.post("", summary="创建性能测试任务")
async def create_task(
    payload: TaskCreate,
    user: User = Depends(require_auth),
):
    result = _service.create_task(
        name=payload.name,
        target_url=payload.target_url,
        method=payload.method,
        test_type=payload.test_type,
        headers=payload.headers,
        body=payload.body,
        business_volume=payload.business_volume,
        user_id=user.id,
        created_by=user.id,
    )
    return Response(data=result)


@router.get("/list", summary="任务列表")
async def list_tasks(
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    test_type: Optional[str] = Query(default=None),
    status: Optional[str] = Query(default=None),
    user: User = Depends(require_auth),
):
    result = _service.list_tasks(
        page=page, page_size=page_size,
        test_type=test_type, status=status,
        user_id=user.id,
    )
    return Response(data=result)


@router.get("/{task_id}", summary="任务详情")
async def get_task(task_id: int, user: User = Depends(require_auth)):
    result = _service.get_task(task_id)
    if not result:
        return Response(code=404, message="任务不存在")
    return Response(data=result)


@router.delete("/{task_id}", summary="删除任务")
async def delete_task(task_id: int, user: User = Depends(require_auth)):
    ok = _service.delete_task(task_id)
    if not ok:
        return Response(code=404, message="任务不存在")
    return Response(message="删除成功")


# ================================================================
# Agent 编排
# ================================================================

@router.post("/{task_id}/plan", summary="生成性能测试方案")
async def run_plan(task_id: int, user: User = Depends(require_auth)):
    result = await _service.run_plan(task_id)
    return Response(data=result)


@router.post("/{task_id}/script", summary="生成测试脚本")
async def run_script(
    task_id: int,
    payload: ScriptGenerate,
    user: User = Depends(require_auth),
):
    result = await _service.run_script(task_id, script_type=payload.script_type)
    return Response(data=result)


@router.post("/{task_id}/analyze", summary="分析性能结果")
async def run_analysis(
    task_id: int,
    payload: AnalyzeRequest = None,
    user: User = Depends(require_auth),
):
    logs = payload.logs if payload else None
    result = await _service.run_analysis(task_id, logs=logs)
    return Response(data=result)


# ================================================================
# 性能诊断 (PerformanceDiagnosticAgent)
# ================================================================

class DiagnoseRequest(BaseModel):
    jstack_dump: str = Field(default="", description="jstack 命令输出文本")
    logs: Optional[List[Any]] = Field(default=None, description="应用日志列表")


@router.post("/{task_id}/diagnose", summary="性能诊断 (jstack/日志/监控分析)")
async def run_diagnostic(
    task_id: int,
    payload: DiagnoseRequest = None,
    user: User = Depends(require_auth),
):
    """执行性能诊断

    输入 jstack dump + 应用日志, 结合监控数据进行深度问题定位:
      - 线程阻塞 / 死锁 / CPU 热点 (jstack 分析)
      - 错误聚类 / 慢操作 / 错误突增 (日志分析)
      - 多源关联分析 (jstack + 日志 + 监控)
      - LLM 根因定位和修复建议
    """
    jstack_dump = payload.jstack_dump if payload else ""
    logs = payload.logs if payload else None
    result = await _service.run_diagnostic(task_id, jstack_dump=jstack_dump, logs=logs)
    if result.get("status") == "error":
        return Response(code=400, message=result.get("error", "诊断失败"))
    return Response(data=result)


# ================================================================
# 测试执行 (PerformanceExecutor)
# ================================================================

@router.post("/{task_id}/execute", summary="启动性能测试执行")
async def run_execute(task_id: int, user: User = Depends(require_auth)):
    """启动 Locust 性能测试执行, 实时采集 TPS/RT/CPU/Memory 指标"""
    result = await _service.run_execute(task_id)
    if result.get("status") == "error":
        return Response(code=400, message=result.get("error", "启动失败"))
    return Response(data=result)


@router.post("/{task_id}/execute/stop", summary="停止性能测试执行")
async def stop_execute(task_id: int, user: User = Depends(require_auth)):
    """停止正在运行的性能测试"""
    result = await _service.stop_execute(task_id)
    return Response(data=result)


@router.get("/{task_id}/execute/status", summary="获取执行状态")
async def get_execute_status(task_id: int, user: User = Depends(require_auth)):
    """获取性能测试执行状态"""
    result = _service.get_execution_status(task_id)
    return Response(data=result)


@router.get("/{task_id}/execute/stream", summary="实时指标 SSE 流")
async def stream_execute_metrics(task_id: int, user: User = Depends(require_auth)):
    """SSE 流式推送实时性能指标

    事件格式 (data: JSON):
      - {"type": "metric", "tps": ..., "avg_rt": ..., "cpu_percent": ..., ...}
      - {"type": "final", "data": {...最终统计...}}
      - {"type": "end", "status": "completed/failed/stopped"}
    """
    async def event_generator():
        async for event in _service.stream_metrics(task_id):
            yield f"data: {json.dumps(event, ensure_ascii=False, default=str)}\n\n"

    return StreamingResponse(
        event_generator(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",
        },
    )


# ================================================================
# 结果和指标
# ================================================================

@router.get("/{task_id}/results", summary="执行结果列表")
async def get_results(task_id: int, user: User = Depends(require_auth)):
    results = _service.get_results(task_id)
    return Response(data=results)


@router.get("/{result_id}/metrics", summary="实时指标")
async def get_metrics(result_id: int, user: User = Depends(require_auth)):
    metrics = _service.get_metrics(result_id)
    return Response(data=metrics)


@router.post("/{task_id}/results", summary="保存执行结果")
async def save_result(
    task_id: int,
    payload: ResultSave,
    user: User = Depends(require_auth),
):
    result = _service.save_result(task_id, payload.model_dump())
    return Response(data=result)


@router.post("/{result_id}/metrics", summary="保存实时指标")
async def save_metric(
    result_id: int,
    payload: MetricSave,
    user: User = Depends(require_auth),
):
    result = _service.save_metric(result_id, payload.model_dump())
    return Response(data=result)
